### PR TITLE
[Scheduler] Cancel pending tasks when the scheduling class becomes infeasible

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

The only reason a task becomes infeasible in local task manager is because of it's scheduling on a placement group, and the placement group has been removed. 

In this case, we should reply to the client that this task/work is cancelled due to pg removal.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
Nice! Can you add tests?

File: src/ray/raylet/scheduling/local_task_manager.cc [hunk 1]
```diff
@@ -750,6 +750,24 @@ bool LocalTaskManager::CancelTask(
...
+      continue;
+    }
+    RAY_LOG(DEBUG) << "Canceling infeasible task "
+                   << (*work_it)->task.GetTaskSpecification().TaskId();
+    ReplyCancelled(
+        *work_it,
+        rpc::RequestWorkerLeaseReply::SCHEDULING_CANCELLED_PLACEMENT_GROUP_REMOVED,
+        "Task/Actor cancelled, the related placement group has been removed.");
+    (*work_it)->SetStateCancelled();
+    tasks.erase(work_it);
```
Comment (wumuzi520) [inline src/ray/raylet/scheduling/local_task_manager.cc:767]: The following code maybe case invalid iterator.
```
for (auto work_it = tasks.begin(); work_it != tasks.end(); work_it++) {
    ...
    tasks.erase(work_it);
}
```

**Review by wumuzi520 — CHANGES_REQUESTED** [review]

File: src/ray/raylet/scheduling/local_task_manager.cc [hunk 2]
```diff
@@ -274,10 +274,10 @@ void LocalTaskManager::DispatchScheduledTasksToWorkers() {
       info_by_sched_cls_.erase(scheduling_class);
     }
     if (is_infeasible) {
-      // TODO(scv119): fail the request.
-      // Call CancelTask
-      tasks_to_dispatch_.erase(shapes_it++);
-    } else if (dispatch_queue.empty()) {
+      CancelInfeasiblePlacementGroupTasks(dispatch_queue);
```
Comment (wumuzi520) [inline src/ray/raylet/scheduling/local_task_manager.cc:277]: It will affect the `work_it` of the current scope and make `work_it` invalid.

**Comment by rkooo567** [comment]
Right now, we have multiple conditions pg tasks can become infeasible.

1. Submit tasks "after" pg is removed
2. pg is removed while tasks are queued.

I think this will handle 2, but 1 will still not be covered. Is there any idea to fix it btw?

File: src/ray/raylet/scheduling/local_task_manager.cc (hunk 2 above)
Comment (scv119) [reply src/ray/raylet/scheduling/local_task_manager.cc:277]: do you mean shape_it?

**Comment by scv119** [comment]
hmm this is a bit tricky to test.

File: src/ray/raylet/scheduling/local_task_manager.cc (hunk 2 above)
Comment (wumuzi520) [reply src/ray/raylet/scheduling/local_task_manager.cc:277]: Oh, sorry, I misread it. This line looks good.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Comment by scv119** [comment]
> Right now, we have multiple conditions pg tasks can become infeasible.
> 
> Submit tasks "after" pg is removed
> pg is removed while tasks are queued.

I think for the first case we need think a little bit more. Maybe some versioning scheme is the way to solve that.

**Comment by scv119** [comment]
cc @rkooo567 @jjyao @iycheng take another look?

**Review by jjyao — COMMENTED** [review]
LGTM

File: src/ray/raylet/local_task_manager.cc [hunk 3]
```diff
@@ -281,10 +281,10 @@ void LocalTaskManager::DispatchScheduledTasksToWorkers() {
       info_by_sched_cls_.erase(scheduling_class);
     }
     if (is_infeasible) {
-      // TODO(scv119): fail the request.
-      // Call CancelTask
-      tasks_to_dispatch_.erase(shapes_it++);
-    } else if (dispatch_queue.empty()) {
+      CancelInfeasiblePlacementGroupTasks(dispatch_queue);
```
[the remark this answers is not in the document]
Comment (scv119) [reply src/ray/raylet/local_task_manager.cc:284]: yeah that makes sense

File: src/ray/raylet/local_task_manager.cc [hunk 4]
```diff
@@ -778,6 +781,31 @@ bool LocalTaskManager::CancelTask(
...
 
+void LocalTaskManager::CancelInfeasiblePlacementGroupTasks(
+    std::deque<std::shared_ptr<internal::Work>> &tasks) {
+  for (auto work_it = tasks.begin(); work_it != tasks.end();) {
+    if ((*work_it)->GetState() != internal::WorkStatus::WAITING) {
+      // Only cancel waiting tasks.
+      ++work_it;
+      continue;
+    }
+    if ((*work_it)->task.GetTaskSpecification().PlacementGroupBundleId().first ==
```
Comment (jjyao) [inline src/ray/raylet/local_task_manager.cc:792]: I think we should move this up, right before calling `CancelInfeasiblePlacementGroupTasks()`. We group tasks by scheduling class, so it's all placement group tasks or none placement group tasks?

Also I think it should be a CHECK to reflect that it's the only cause now:

I think it should be something like
```
if (infeasible) {
CHECK(worker->IsplacementGroupTask());
CancelInfeasiblePlacementGroupTasks();
}
```

File: src/ray/raylet/local_task_manager.cc [hunk 5]
```diff
@@ -778,6 +781,31 @@ bool LocalTaskManager::CancelTask(
   return false;
 }
 
+void LocalTaskManager::CancelInfeasiblePlacementGroupTasks(
+    std::deque<std::shared_ptr<internal::Work>> &tasks) {
+  for (auto work_it = tasks.begin(); work_it != tasks.end();) {
+    if ((*work_it)->GetState() != internal::WorkStatus::WAITING) {
```
Comment (jjyao) [inline src/ray/raylet/local_task_manager.cc:787]: Hmm, what will happen to other WorkStatus like WAIINT_FOR_WORKER when a pg is removed.

**Pull request closed without merging**